BUG: Do not round DatetimeIndex nanosecond precision when iterating

mroeschke opened the pull request:
- [x] closes #19603
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

~~This issue was actually very specific to localizing a `DatetimeIndex` with `datetime.timezone.utc`~~

```
# On master
In [1]: import pandas as pd

In [2]: datetimeindex = pd.DatetimeIndex(["2018-02-08 15:00:00.168456358"], tz='UTC')

In [3]: list(datetimeindex)[0] == datetimeindex[0] # returned False with datetime.timezone.utc
Out[3]: True 
```

~~I realize our timezone support relies heavily on pytz and dateutil, but I am curious how much support we have for `datetime.timezone` objects.~~

So the actual issue was that `datetime.timezone.utc` is (rightly) considered a fixed offset, and the code path was calculating the new value with Python `datetimes` and `timedeltas` which doesn't yet support nanosecond resolution.

mroeschke committed:
Bug: Do not round DatetimeIndex nanosecond precision when iterating

Fix the issue within fix_offsets

add addtional dateutil test

Address_edits

jreback commented on pandas/_libs/tslibs/timezones.pyx:
```diff
@@ -30,6 +32,10 @@ cdef int64_t NPY_NAT = get_nat()
 # ----------------------------------------------------------------------
 
 cdef inline bint is_utc(object tz):
+    if PY3:
```
this would be a perf issue, you can almost certainly access this in stead at the cython level

jreback commented on pandas/tests/indexes/datetimes/test_datetime.py:
```diff
@@ -7,12 +7,18 @@
 from pandas.util.testing import assert_almost_equal
 
+if PY3:
+    from datetime import timezone
```
this is not idiomatic, see comments above

jreback replied:
rather than put this here, make a fixture in pandas/conftest.py

called datetime_tz_utc to return this and skip if its not PY3

jreback reviewed (changes requested):
we would need comprehensive support for datetime.timezone in order to accept this. IOW you need to add testing in LOTS of places where we test datetutil/pytz

mroeschke commented:
I updated the fix and found the real issue (edited in my description above).

jreback commented on pandas/_libs/tslib.pyx:
```diff
@@ -146,10 +146,12 @@ def ints_to_pydatetime(ndarray[int64_t] arr, tz=None, freq=None,
-                    if box:
-                        dt = Timestamp(dt)
-                    result[i] = dt
+                    # Python datetime objects do not support nanosecond
```
use tz_convert_tzlocal_to_utc here from conversion.pyx

jreback commented on pandas/tests/indexes/datetimes/test_datetime.py:
```diff
@@ -258,6 +264,13 @@ def test_iteration_preserves_tz(self):
 
+    @pytest.mark.parametrize('tz', [None, 'UTC', "US/Central", dt_UTC,
+                                    dateutil.tz.tzoffset(None, -28800)])
+    def test_iteration_preserves_nanoseconds(self, tz):
```
add a similar test for pandas/tests/scalar/test_timezones.py

jreback commented on pandas/tests/indexes/datetimes/test_datetime.py:
```diff
@@ -258,6 +264,13 @@ def test_iteration_preserves_tz(self):
             assert result == expected
 
+    @pytest.mark.parametrize('tz', [None, 'UTC', "US/Central", dt_UTC,
+                                    dateutil.tz.tzoffset(None, -28800)])
```
move to test_timezones (in same subdir)

jreback commented on pandas/tests/indexes/datetimes/test_datetime.py:
```diff
@@ -258,6 +264,13 @@ def test_iteration_preserves_tz(self):
+    def test_iteration_preserves_nanoseconds(self, tz):
+        # GH 19603
+        index = pd.DatetimeIndex(["2018-02-08 15:00:00.168456358"], tz=tz)
+        assert list(index)[0] == index[0]
```
check both getitem and iteration

mroeschke commented:
Replaced fix with the use of `tz_convert_utc_to_tzlocal`, added `datetime_tz_utc` fixture, Also I am not sure what test I should have added to `pandas/tests/scalar/test_timezones.py` since the issue is with `DatetimeIndex.__iter__`

mroeschke committed:
Fix fixture

mroeschke commented:
All green

jreback commented:
thanks @mroeschke 

FYI looking to move tz to be fixtures (e.g. the strings 'UTC', 'Asia/Tokyo') etc....across the test codebase, IOW to have a 'tz' fixture (prob need several slightly different ones)

cc @jbrockmendel
